### PR TITLE
Feat: API Features

### LINKED ISSUE #20: Add filtering to get all

### PR DESCRIPTION
**Pull request by Xaroz**
Closes #20 

- Added `filtering`, `sorting`,  `selecting` and `paginating` to get all function
- Added a way to get `totalDocuments` count from get all
- Moved Visibility to `types/request.ts`
- Added example `.env` file

### CODE REVIEW

**Review by Bant89 — APPROVED** [review]
Excellent work!

File: src/controllers/handlerFactory.ts [hunk 1]
```diff
@@ -161,12 +155,23 @@ export const getAllEntities = <ModelInterface>(
...
+
+      const documents: ModelInterface[] = await features.query;
+      const totalDocuments = await ModelEntity.countDocuments(
+        features.query.getFilter()
+      );
 
       res.status(200).json({
         status: 'success',
         results: documents.length,
+        totalDocuments,
```
Comment (bujosa) [inline src/controllers/handlerFactory.ts:174]: That totalDocument comes in json format? total likes: 5? or count 

if it is not the case, you can put it total or count : totalDocuments.

**Review by bujosa — APPROVED** [review]
LGTM

File: src/utils/apiFeatures.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  }
+
+  filter() {
+    const queryObject = { ...this.queryString };
+
+    const excludedFields = ['page', 'sort', 'limit', 'fields'];
+    excludedFields.forEach((el) => delete queryObject[el]);
+
+    let queryStr = JSON.stringify(queryObject);
+    queryStr = queryStr.replace(/\b(gte|gt|lte|lt)\b/g, (match) => `$${match}`);
```
Comment (xeno097) [inline src/utils/apiFeatures.ts:28]: cool 🔥

File: src/utils/apiFeatures.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    const page = this.queryString.page ? parseInt(this.queryString.page) : 1;
+    const limit = this.queryString.limit
+      ? parseInt(this.queryString.limit)
+      : 20;
+    const skip = (page - 1) * limit;
+
+    this.query = this.query.skip(skip).limit(limit);
+
+    return this;
+  }
```
Comment (xeno097) [inline src/utils/apiFeatures.ts:65]: Just for reference for the future https://jsonapi.org/profiles/ethanresnick/cursor-pagination/#:~:text=Cursor%2Dbased%20pagination%20(aka%20keyset,be%20shifted%20forward%20by%20one.

**Review by xeno097 — APPROVED** [review]
Nicely done 🔥

**Pull request merged**